lnwire/dyn_propose.go

```diff
@@ -0,0 +1,316 @@
+package lnwire
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// DPDustLimitSatoshis is the TLV type number that identifies the record
+	// for DynPropose.DustLimit.
+	DPDustLimitSatoshis tlv.Type = 0
+
+	// DPMaxHtlcValueInFlightMsat is the TLV type number that identifies the
+	// record for DynPropose.MaxValueInFlight.
+	DPMaxHtlcValueInFlightMsat tlv.Type = 1
+
+	// DPChannelReserveSatoshis is the TLV type number that identifies the
+	// for DynPropose.ChannelReserve.
+	DPChannelReserveSatoshis tlv.Type = 2
+
+	// DPToSelfDelay is the TLV type number that identifies the record for
+	// DynPropose.CsvDelay.
+	DPToSelfDelay tlv.Type = 3
+
+	// DPMaxAcceptedHtlcs is the TLV type number that identifies the record
+	// for DynPropose.MaxAcceptedHTLCs.
+	DPMaxAcceptedHtlcs tlv.Type = 4
+
+	// DPFundingPubkey is the TLV type number that identifies the record for
+	// DynPropose.FundingKey.
+	DPFundingPubkey tlv.Type = 5
+
+	// DPChannelType is the TLV type number that identifies the record for
+	// DynPropose.ChannelType.
+	DPChannelType tlv.Type = 6
+
+	// DPKickoffFeerate is the TLV type number that identifies the record
+	// for DynPropose.KickoffFeerate.
+	DPKickoffFeerate tlv.Type = 7
+)
+
+// DynPropose is a message that is sent during a dynamic commitments negotiation
+// process. It is sent by both parties to propose new channel parameters.
+type DynPropose struct {
+	// ChanID identifies the channel whose parameters we are trying to
+	// re-negotiate.
+	ChanID ChannelID
+
+	// Initiator is a byte that identifies whether this message was sent as
+	// the initiator of a dynamic commitment negotiation or the responder
+	// of a dynamic commitment negotiation. bool true indicates it is the
+	// initiator
+	Initiator bool
```

Comment:
Isn't this information already apparent to both sides contextually? 

Eg: for channel funding we know who initiator is from the very start, so we don't need messages like `closing_signed` to signal if it's sent by the initiator not (for that msg behavior differs based on if you're the initiator or not).

Reply:
What do we do in the situation where both sides send a `dyn_propose` "at the same time"? It's unlikely in practice but possible. I suppose we could just fail at some point when the conflict becomes unignorable. This was put in to express intent to the receiver that we believe we are initiating the negotiation. Such an intent isn't clear otherwise.

Reply:
I think this can be fixed if you abort the flow if you send a `dyn_propose` with `initiator=true` and receive a `dyn_propose` with `initiator=true`

Reply:
Agreed. But absent the initiator=true label I don't think this can be accomplished.